[libc++] Add support for picolibc and newlib in RUNTIMES_USE_LIBC

voltur01 opened the pull request:
This replaces detection of picolibc in libc++ (by checking for and including picolibc.h) with using RUNTIMES_USE_LIBC build time option intriduced in https://github.com/llvm/llvm-project/pull/134893 RUNTIMES_USE_LIBC is extended to accept picolibc and newlib.

Detection of picolibc via the header is kept as a deprecated feature to avoid breaking builds.

libc++ is updated to use dedicated LIBCXX_LIBC_NEWLIB macro to check for newlib specific conditions instead of less informative _NEWLIB_VERSION

voltur01 committed:
[libc++] Add support for picolibc and newlib in RUNTIMES_USE_LIBC

This replaces detection of picolibc in libc++ (by checking for and including picolibc.h) with using RUNTIMES_USE_LIBC build time option intriduced in https://github.com/llvm/llvm-project/pull/134893 RUNTIMES_USE_LIBC is extended to accept picolibc and newlib.

Detection of picolibc via the header is kept as a deprecated feature to avoid breaking builds.

libc++ is updated to use dedicated LIBCXX_LIBC_NEWLIB macro to check for newlib specific conditions instead of less informative _NEWLIB_VERSION

pratlucas commented on libcxx/include/__configuration/platform.h:
```diff
@@ -42,11 +42,10 @@
+// To build libc++ with picolibc provide RUNTIMES_USE_LIBC=picolibc
 #if __has_include(<picolibc.h>)
-#  include <picolibc.h>
+#  define LIBCXX_LIBC_NEWLIB
```
As the plan is to deprecate this use case after LLVM 22, should a deprecation warning message be added here?

pratlucas replied:
Brilliant! Thanks

pratlucas reviewed (commented):
Just a small comment, but LGTM otherwise.

voltur01 committed:
Provide a warning if RUNTIMES_USE_LIBC is not used

pratlucas reviewed (approved):
LGTM, but as per the contribution policy this PR probably needs approval from someone from the libc++ review group.

ldionne commented on libcxx/CMakeLists.txt:
```diff
@@ -481,6 +481,21 @@ include(config-ix)
 include(HandleLibC) # Setup the C library flags
 include(HandleLibCXXABI) # Setup the ABI library flags
 
+# Set C library in use to define respective macro in __config_site
```
I would suggest we do this instead (around line 753 where we handle other `__config_site` variables):

```
if (RUNTIMES_USE_LIBC STREQUAL "llvm-libc")
  config_define(_LIBCPP_LIBC_LLVMLIBC 1)
elseif (RUNTIMES_USE_LIBC STREQUAL "picolibc")
  config_define(_LIBCPP_LIBC_PICOLIBC 1)
elseif (RUNTIMES_USE_LIBC STREQUAL "newlib")
  config_define(_LIBCPP_LIBC_NEWLIB 1)
endif()
```

That's more consistent with what we do for the other `__config_site` macros.

ldionne commented on libcxx/CMakeLists.txt:
```diff
@@ -481,6 +481,21 @@ include(config-ix)
+  set(LIBCXX_LIBC_LLVMLIBC 1)
+elseif (RUNTIMES_USE_LIBC STREQUAL "picolibc")
+  set(LIBCXX_LIBC_PICOLIBC 1)
+  # picolibc is derived from newlib and behaves the same in regards to libc++
+  # so setting both here:
+  # * LIBCXX_LIBC_NEWLIB is used now
+  # * LIBCXX_LIBC_PICOLIBC can be used for further customizations later
+  set(LIBCXX_LIBC_NEWLIB 1) 
```
Do you absolutely need this right now (if so, why)? I think it would be cleaner to have a 1-1 mapping between the macros we define and the libc we're using, otherwise things become confusing when reading the source.

ldionne commented on libcxx/CMakeLists.txt:
```diff
@@ -481,6 +481,21 @@ include(config-ix)
 include(HandleLibC) # Setup the C library flags
```
General comment on the whole patch: Let's use `_LIBCPP_foo` instead of `LIBCXX_foo` for the macro names. That's what we do consistently everywhere, and we must use a reserved identifier.

ldionne commented on libcxx/include/__config_site.in:
```diff
@@ -42,6 +42,11 @@
 #cmakedefine _LIBCPP_HARDENING_MODE_DEFAULT @_LIBCPP_HARDENING_MODE_DEFAULT@
 
+// C libraries
+#cmakedefine LIBCXX_LIBC_LLVMLIBC
+#cmakedefine LIBCXX_LIBC_PICOLIBC
+#cmakedefine LIBCXX_LIBC_NEWLIB
```
I'd use `#cmakedefine01` here and then I'd check `#if _LIBCPP_LIBC_NEWLIB` instead of `#if defined(_LIBCPP_LIBC_NEWLIB)`. This can prevent mistakes when coupled with `-Wundef`.

voltur01 replied:
Very good suggestion, done - the checks are much cleaner now.

ldionne commented on libcxx/include/__configuration/platform.h:
```diff
@@ -42,11 +42,15 @@
-//       user code. Move code paths that need _NEWLIB_VERSION to another customization mechanism.
-#if __has_include(<picolibc.h>)
-#  include <picolibc.h>
+// TODO: Remove this deprecated behavior after LLVM 22 release
```
Since this is something that vendors should change, I'm not sure there's a benefit from deprecating this before we remove it.

voltur01 replied:
Agreed, I do not expect many people using it anyway - removed.

ldionne commented on libcxx/include/__cxx03/__fwd/ios.h:
We shouldn't modify the `__cxx03` headers. However, that does cause a problem since the `__config_site` is shared with the frozen C++03 headers. @philnik777, do you have thoughts on what to do here?

philnik777 replied:
I think this is a reasonable enough change.

ldionne commented on runtimes/cmake/Modules/HandleLibC.cmake:
```diff
@@ -10,10 +10,10 @@
 set(RUNTIMES_USE_LIBC "system" CACHE STRING "Specify C library to use. Supported values are ${RUNTIMES_SUPPORTED_C_LIBRARIES}.")
 if (NOT "${RUNTIMES_USE_LIBC}" IN_LIST RUNTIMES_SUPPORTED_C_LIBRARIES)
-  message(FATAL_ERROR "Unsupported C library: '${RUNTIMES_CXX_ABI}'. Supported values are ${RUNTIMES_SUPPORTED_C_LIBRARIES}.")
+  message(FATAL_ERROR "Unsupported C library: '${RUNTIMES_USE_LIBC}'. Supported values are ${RUNTIMES_SUPPORTED_C_LIBRARIES}.")
```
Good catch!

ldionne reviewed (changes requested):
Thanks for the patch! I have some comments but this makes sense to me.

voltur01 committed:
Address review comments

- Use _LIBCXX_... reserved namespace macro names.
- Move library definition code closer to the rest of __config_site definitions in CMake.
- Remove unused LIBCXX_LIBC_NEWLIB definition.
- Use #cmakedefine01 and thus avoid defined() in conditional statements.
- Remove deprecation check.

arichardson commented:
It looks to me like this will break existing builds that don't define the macro. I think for at least a while there should be a CMake check for picoline.h (and the newlib equivalent) to set these macros correctly and emit a warning to update the build config?

voltur01 commented:
Thank you for the review @arichardson! There is a way to provide a warning like in this commit https://github.com/llvm/llvm-project/pull/147956/commits/a8769880f65faf2abce6bee04b480efd5c0bf571 however I am not sure how many people are actually building with picolibc and will benefit from it.

I think this is up to maintainers to decide if the warning should be included or not.

arichardson commented:
> Thank you for the review @arichardson! There is a way to provide a warning like in this commit [a876988](https://github.com/llvm/llvm-project/commit/a8769880f65faf2abce6bee04b480efd5c0bf571) however I am not sure how many people are actually building with picolibc and will benefit from it.
> 
> I think this is up to maintainers to decide if the warning should be included or not.

Yeah I'm not sure how important this is. I am using the existing implicit support but I can easily update my build scripts.
I think it would be quite nice to detect this at CMake time and give a warning, but probably no need to add the checks in the headers themselves.
IMO this is a question for the libc++ maintainers

saturn691 commented:
This PR would be very useful to me as an LLVM-libc contributor, if you could add back support for `_LIBCPP_LIBC_LLVMLIBC ` - I would like a define that I can access through `libcxx/include/__config`- currently the `__LLVM_LIBC__` define only works during compile time (of the toolchain) and doesn't work during compile time (of the user code).

voltur01 committed:
Rename _LIBCXX_* to _LIBCPP_*

Rename _LIBCXX_* options to _LIBCPP_* to follow libcxx conventions.

ldionne committed:
Merge branch 'main' into newlib_support

ldionne reviewed (approved):
> I think the only outstanding question above was if we needed to provide compatibility with the old way and a warning.

I am personally neutral on that check. I'm fine with your current patch.

LGTM as-is assuming the CI is green. Apologies for the review delay.

s-barannikov commented:
> Do you see any failures or other issues related to the checks?

No, just noticed that there are a couple more places where the new option/define could be used.
